### PR TITLE
Create a function to get notebook python version for execution

### PR DESCRIPTION
**Pull request by soheilazangeneh**
Notebooks might require a specific python version for execution. This PR creates a function that parses the notebook and gets the python version. Then passes that to the yaml file for execution.

### CODE REVIEW

File: .cloud-build/execute_changed_notebooks_helper.py [hunk 1]
```diff
@@ -447,25 +447,5 @@ def process_and_execute_notebooks(
         # Raise error if any notebooks failed
         if not all([result.is_pass for result in results_sorted]):
             raise RuntimeError("Notebook failures detected. See logs for details")
-
```
Comment (soheilazangeneh) [inline .cloud-build/execute_changed_notebooks_helper.py:436]: Removed this part for 2 main reason:
1. Adding another condition to the if statement at line 356 for python versions might make sense for now, but if we want to add more features in future we have to make that if statement conditions longer and longer. This way, every notebook is ran the same way no matter how many notebooks are found.
2. Better readability. First, when I read the code I was confused that we have 2 different process based on the fact that we have 1 notebook and more than 1. I think a separate process for want notebook makes changes more complicated.

**Review by soheilazangeneh — COMMENTED** [review]
/gcbrun

File: .cloud-build/execute_changed_notebooks_helper.py [hunk 2]
```diff
@@ -110,6 +114,31 @@ def _process_notebook(
...
+    file = open(notebook_path) # Open the ipynb file that is passed in
+    src = file.read() # Read the file in
+    j = json.loads(src) # Parse the JSON into an object
+
+    for cell in j['cells']: #Iterate over all the cells in the ipynb
+        if cell['cell_type'] == 'markdown':
+            markdown = str.join('', cell['source'])
+
+            # Look for the python version specification pattern
+            re_match = re.search('python\ ?version[:,=]\ ?\d(.\d)?', markdown, flags=re.IGNORECASE)
```
Comment (ivanmkc) [inline .cloud-build/execute_changed_notebooks_helper.py:133]: Strange to use regex twice like this. Can't you just extract using the first regex?

File: .cloud-build/execute_changed_notebooks_helper.py [hunk 3]
```diff
@@ -110,6 +114,31 @@ def _process_notebook(
...
+def _get_notebook_python_version(notebook_path: str) -> str:
+    """
+    Get the python version for running the notebook if it is specified in
+    the notebook.
+    """
+    python_version = PYTHON_VERSION
+
+    file = open(notebook_path) # Open the ipynb file that is passed in
+    src = file.read() # Read the file in
+    j = json.loads(src) # Parse the JSON into an object
```
Comment (ivanmkc) [inline .cloud-build/execute_changed_notebooks_helper.py:126]: Use a more description variable name.

File: .cloud-build/execute_changed_notebooks_helper.py (hunk 1 above)
Comment (ivanmkc) [reply .cloud-build/execute_changed_notebooks_helper.py:436]: The separate process for a single notebook is needed as it's the only way to show results in the public logs (without having users go through a secondary download of executed notebooks).

Feel free to refactor but we need that option which as 99% of use cases.

File: .cloud-build/execute_notebook_remote.py [hunk 4]
```diff
@@ -50,8 +51,12 @@ def execute_notebook_remote(
         "_PYTHON_IMAGE": container_uri,
         "_NOTEBOOK_GCS_URI": notebook_uri,
         "_NOTEBOOK_OUTPUT_GCS_URI": notebook_output_uri,
+        "_PYTHON_VERSION" : "python" + python_version
```
Comment (ivanmkc) [inline .cloud-build/execute_notebook_remote.py:54]: Prefer f-string: `f"python{python_version}"`

File: .cloud-build/notebook-execution-test-cloudbuild-single.yaml [hunk 5]
```diff
@@ -10,29 +10,29 @@ steps:
...
+    - ${_PYTHON_VERSION} -m venv workspace/env
   # Install Python dependencies
   - name: ${_PYTHON_IMAGE}
     entrypoint: /bin/sh
     args:
     - -c
     - . workspace/env/bin/activate &&
-      python3 -m pip -q install -U pip &&
-      python3 -m pip -q install -U -r .cloud-build/requirements.txt
+      ${_PYTHON_VERSION} -m pip -q install -U pip &&
```
Comment (ivanmkc) [inline .cloud-build/notebook-execution-test-cloudbuild-single.yaml:26]: I actually don't think you need this part after calling `workspace/env/bin/activate`.

Try using `python -m pip -q install ...`

Comment (ivanmkc) [reply .cloud-build/notebook-execution-test-cloudbuild-single.yaml:26]: The `python` used should be inferred from the env.

File: .cloud-build/notebook-execution-test-cloudbuild-single.yaml [hunk 6]
```diff
@@ -10,29 +10,29 @@ steps:
...
+      ${_PYTHON_VERSION} -m pip -q install -U -r .cloud-build/requirements.txt
   # Install Python dependencies and run testing script
   - name: ${_PYTHON_IMAGE}
     entrypoint: /bin/sh
     args:
     - -c
     - |
       . workspace/env/bin/activate &&
-      python3 .cloud-build/execute_notebook_cli.py --notebook_source "${_NOTEBOOK_GCS_URI}" --output_file_or_uri "${_NOTEBOOK_OUTPUT_GCS_URI}"
+      ${_PYTHON_VERSION} .cloud-build/execute_notebook_cli.py --notebook_source "${_NOTEBOOK_GCS_URI}" --output_file_or_uri "${_NOTEBOOK_OUTPUT_GCS_URI}"
```
Comment (ivanmkc) [inline .cloud-build/notebook-execution-test-cloudbuild-single.yaml:35]: Ditto.

File: .cloud-build/execute_changed_notebooks_helper.py (hunk 3 above)
Comment (soheilazangeneh) [reply .cloud-build/execute_changed_notebooks_helper.py]: Done: e6c2c44ce4c1625d2ac79da9861bf42fc644bcc5

File: .cloud-build/execute_changed_notebooks_helper.py [hunk 7]
```diff
@@ -167,6 +196,10 @@ def process_and_execute_notebook(
     time_start = datetime.datetime.now()
     operation = None
     try:
+        # Get the python version for ruuning the notebook if specified
```
[the remark this answers is not in the document]
Comment (soheilazangeneh) [reply .cloud-build/execute_changed_notebooks_helper.py]: Done: de745db35d871cf3dba3a21fa1e03ec5f898691d

File: .cloud-build/execute_notebook_remote.py (hunk 4 above)
Comment (soheilazangeneh) [reply .cloud-build/execute_notebook_remote.py]: Done: cbb967268f81732c00401a90d236ad8ae04d74db

File: .cloud-build/notebook-execution-test-cloudbuild-single.yaml (hunk 5 above)
Comment (soheilazangeneh) [reply .cloud-build/notebook-execution-test-cloudbuild-single.yaml]: Done: a70b32277fa885e60f3850b8ad294ba763798106

File: .cloud-build/notebook-execution-test-cloudbuild-single.yaml (hunk 6 above)
Comment (soheilazangeneh) [reply .cloud-build/notebook-execution-test-cloudbuild-single.yaml]: Done: a70b32277fa885e60f3850b8ad294ba763798106

File: .cloud-build/execute_changed_notebooks_helper.py (hunk 2 above)
Comment (soheilazangeneh) [reply .cloud-build/execute_changed_notebooks_helper.py]: Done: cf86cf923aa4ed2fde3b91154911c3a3c0035dd1
Thank you for your help! :)

**Comment by soheilazangeneh** [comment]
- [ ] Add a log separator that makes the log more clear
- [ ] Remove extra steps from the child build

File: .cloud-build/execute_changed_notebooks_helper.py [hunk 8]
```diff
@@ -110,6 +115,31 @@ def _process_notebook(
...
 
 
+def _get_notebook_python_version(notebook_path: str) -> str:
+    """
+    Get the python version for running the notebook if it is specified in
+    the notebook.
+    """
+    python_version = PYTHON_VERSION
+
+    file = open(notebook_path) # Open the ipynb file that is passed in
```
Comment (ivanmkc) [inline .cloud-build/execute_changed_notebooks_helper.py:125]: For consistency: Maybe stick with having comments on top of the line instead of to the right

**Review by ivanmkc — APPROVED** [review]
LGTM

File: .cloud-build/execute_changed_notebooks_helper.py (hunk 8 above)
Comment (soheilazangeneh) [reply .cloud-build/execute_changed_notebooks_helper.py]: Done.

**Pull request merged**